firmware/src/boards/cynthion_d11/board_rev.c

```diff
@@ -50,12 +50,20 @@ void detect_hardware_revision(void)
 
     // Retrieve a single ADC reading.
     hri_adc_set_SWTRIG_START_bit(ADC);
-	while (!hri_adc_get_interrupt_RESRDY_bit(ADC));
-	reading = hri_adc_read_RESULT_reg(ADC);
+    while (!hri_adc_get_interrupt_RESRDY_bit(ADC));
+    reading = hri_adc_read_RESULT_reg(ADC);
+
+    /*
+     * Read the ADC a second time and discard the previous reading to eliminate
+     * error. Initial readings at start-up are typically 0.2% to 1.0% high.
+     */
+    hri_adc_set_SWTRIG_START_bit(ADC);
+    while (!hri_adc_get_interrupt_RESRDY_bit(ADC));
+    reading = hri_adc_read_RESULT_reg(ADC);
 
     // Convert ADC measurement to per mille of the reference voltage.
-    uint32_t permille = (((uint32_t)reading * 1000) + 20480) >> 12;
-    if (permille > 510) {
+    uint32_t permille = ((uint32_t)reading * 1000) >> 12;
```

Comment:
Since this is now a floor operation instead of rounding, shouldn't the thresholds be lower rather than higher? Or am I missing something?

Reply:
Maybe you're missing the inversion (`permille = 1000 - permille;`)?

In terms of raw `reading` values, the threshold between r1.3 and r1.4 for `gsg_production` boards was effectively `3175` before this PR and is still `3175` with the PR. `3175` is 77.5% of the maximum ADC reading of `4096`. 77.5% is the designed threshold.

For non-`gsg_production` boards the threshold between r1.3 and r1.4 was effectively `881` previously but is now corrected to `922` with this PR. `922` is 22.5% of `4096`. 22.5% is the designed threshold. Without this PR, prototype non-`gsg_production` r1.4 boards were frequently misidentified as r1.3.